File: ncarg2d/src/libncarg/support/idsfft.f
C
C $Id: idsfft.f,v 1.6 2008-07-27 00:17:30 haley Exp $
C                                                                      
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C The use of this Software is governed by a License Agreement.
C
      SUBROUTINE IDSFFT (MD,NDP,XD,YD,ZD,NXI,NYI,NZI,XI,YI,ZI,IWK,WK)
C
C DIMENSION OF           XD(NDP), YD(NDP), ZD(NDP), XI(NXI),
C ARGUMENTS              YI(NYI), ZI(NZI,NYI), WK(6*NDP),
C                        IWK(31*NDP + NXI*NYI)
C
C PURPOSE                This subroutine performs smooth surface
C                        fitting when the projections of the data
C                        points in the X-Y plane are irregularly
C                        distributed in the plane.
C
C USAGE                  CALL IDSFFT (MD,NDP,XD,YD,ZD,NXI,NYI,NZI,
C                                     XI,YI,ZI,IWK,WK)
C
C ARGUMENTS
C
C ON INPUT               MD
C                          Mode of computation (must be 1, 2, or 3,
C                          else an error return will occur).
C                           = 1 if this is the first call to this
C                               subroutine, or if the value of NDP
C                               has been changed from the previous
C                               call, or if the contents of the XD
C                               or YD arrays have been changed from
C                               the previous call.
C                           = 2 if the values of NDP and the XD and
C                               YD arrays are unchanged from the
C                               previous call, but new values for
C                               XI and YI are being used.  If MD = 2
C                               and NDP has been changed since the
C                               previous call to IDSFFT, an error
C                               return occurs.
C                           = 3 if the values of NDP, NXI, NYI, XD,
C                               YD, XI, and YI are unchanged from the
C                               previous call, i.e., if the only change
C                               on input to IDSFFT is in the ZD array.
C                               If MD = 3 and NDP, NXI, or NYI has been
C                               changed since the previous call to
C                               IDSFFT, an error return occurs.
C
C                           Between the call with MD = 2 or MD = 3 and
C                           the preceding call, the IWK and WK work
C                           arrays should not be disturbed.
C
C                        NDP
C                          Number of data points (must be 4 or
C                          greater, else an error return will occur).
C
C                        XD
C                          Array of dimension NDP containing the X
C                          coordinates of the data points.
C
C                        YD
C                          Array of dimension NDP containing the Y
C                          coordinates of the data points.
C
C                        ZD
C                          Array of dimension NDP containing the Z
C                          coordinates of the data points.
C
C                        NXI
C                          Number of output grid points in the X
C                          direction (must be 1 or greater, else
C                          an error return will occur).
C
C                        NYI
C                          Number of output grid points in the Y
C                          direction (must be 1 or greater, else
C                          an error return will occur).
C
C                        NZI
C                          First dimension of ZI as declared in the
C                          calling program.  NZI must be greater than
C                          or equal to NXI, else an error return will
C                          occur.
C
C                        XI
C                          Array of dimension NXI containing the
C                          X coordinates of the output grid points.
C
C                        YI
C                         Array of dimension NYI containing the
C                         Y coordinates of the output grid points.
C
C                        IWK
C                          Integer work array of dimension at
C                          least 31*NDP + NXI*NYI.
C
C                        WK
C                          Real work array of dimension at least 6*NDP.
C
C ON OUTPUT              ZI
C                           Real, two-dimensional array of dimension
C                           (NZI,NYI), storing the interpolated Z
C                           values at the output grid points.
C
C SPECIAL CONDITIONS     Inadequate work space IWK and WK may cause
C                        incorrect results.
C
C                        The data points must be distinct and their
C                        projections in the X-Y plane must not be
C                        collinear, else an error return occurs.
C
C IDSFFT calls the subroutines IDGRID, IDPDRV, IDPTIP, and IDTANG.
C
C Declaration statements.
C
      DIMENSION XD(NDP),YD(NDP),ZD(NDP),XI(NXI),YI(NYI),ZI(NZI,NYI),
     +          IWK(31*NDP+NXI*NYI),WK(6*NDP)
C
      COMMON /IDPT/ ITPV,DMMY(27)
      SAVE   /IDPT/
C
      COMMON /IDCOMN/ INTY,ITTY,ALSP,BLSP,CLSP,XAVG,YAVG
      SAVE   /IDCOMN/
C
C Check for an uncleared prior error.
C
        IF (ICFELL('IDSFFT (BIVAR) - UNCLEARED PRIOR ERROR',1).NE.0)
     +                                                        RETURN
C
C Check for input errors.
C
      IF (MD.LT.1.OR.MD.GT.3) THEN
        CALL SETER ('IDSFFT (BIVAR) - INPUT VARIABLE MD IS OUT OF RANGE'
     +,2,1)
        RETURN
      END IF
C
      IF (NDP.LT.4) THEN
        CALL SETER ('IDSFFT (BIVAR) - INPUT VARIABLE NDP IS OUT OF RANGE
     +',3,1)
        RETURN
      END IF
C
      IF (NXI.LT.1.OR.NYI.LT.1) THEN
        CALL SETER ('IDSFFT (BIVAR) - INPUT VARIABLE NXI OR NYI IS OUT O
     +F RANGE',4,1)
        RETURN
      END IF
C
      IF (NXI.GT.NZI) THEN
        CALL SETER ('IDSFFT (BIVAR) - INPUT VARIABLE NZI IS LESS THAN NX
     +I',5,1)
        RETURN
      END IF
C
      IF (MD.LE.1) THEN
C
        IWK(1)=NDP
C
      ELSE
C
        NDPPV=IWK(1)
C
        IF (NDP.NE.NDPPV) THEN
          CALL SETER  ('IDSFFT (BIVAR) - MD = 2 OR 3 BUT NDP WAS CHANGED
     + SINCE LAST CALL',6,1)
          RETURN
        END IF
C
      END IF
C
      IF (MD.LE.2) THEN
C
        IWK(2)=INTY
        IWK(3)=NXI
        IWK(4)=NYI
C
      ELSE
C
        INTYPV=IWK(2)
C
        IF (INTY.NE.INTYPV) THEN
          CALL SETER ('IDSFFT (BIVAR) - MD = 3 BUT ITY WAS CHANGED SINCE
     + LAST CALL',7,1)
           RETURN
        END IF
C
        NXIPV=IWK(3)
C
        IF (NXI.NE.NXIPV) THEN
          CALL SETER ('IDSFFT (BIVAR) - MD = 3 BUT NXI WAS CHANGED SINCE
     + LAST CALL',8,1)
           RETURN
        END IF
C
        NYIPV=IWK(4)
C
        IF (NYI.NE.NYIPV) THEN
          CALL SETER ('IDSFFT (BIVAR) - MD = 3 BUT NYI WAS CHANGED SINCE
     + LAST CALL',9,1)
          RETURN
        END IF
C
      END IF
C
C Allocate storage areas in the array IWK.
C
      JWIPT=16
      JWIWL=6*NDP+1
      JWNGP0=JWIWL-1
      JWIPL=24*NDP+1
      JWIWP=30*NDP+1
      JWIGP0=31*NDP
      JWWPD=5*NDP+1
C
C If MD = 1, triangulate the X/Y plane.  (If MD = 2 or 3, this has
C already been done and need not been redone.)
C
      IF (MD.EQ.1) THEN
        CALL IDTANG (NDP,XD,YD,NT,IWK(JWIPT),NL,IWK(JWIPL),IWK(JWIWL),
     +                                                  IWK(JWIWP),WK)
        IF (ICFELL('IDSFFT',10).NE.0) RETURN
        IWK(5)=NT
        IWK(6)=NL
      ELSE
        NT=IWK(5)
        NL=IWK(6)
      END IF
C
      IF (NT.EQ.0) RETURN
C
C If linear interpolation is activated, compute the coefficients of the
C least-squares-fit plane and the mean values of X and Y.
C
      IF (INTY.NE.0) THEN
        CALL IDLSQF (XD,YD,ZD,NDP,ALSP,BLSP,CLSP,XAVG,YAVG)
      END IF
C
C If MD = 1 or 2, sort output grid points in ascending order by the
C number of the containing region (a triangle number or a combined
C pair of border line segment numbers).  (If MD = 3, this has already
C been done and need not be redone.)
C
      IF (MD.LE.2) THEN
        CALL IDGRID (XD,YD,NT,IWK(JWIPT),NL,IWK(JWIPL),NXI,
     +               NYI,XI,YI,IWK(JWNGP0+1),IWK(JWIGP0+1))
        IF (ICFELL('IDSFFT',11).NE.0) RETURN
      END IF
C
C If quintic interpolation is activated, estimate partial derivatives.
C
      IF (INTY.EQ.0) THEN
        CALL IDPDRV (NDP,XD,YD,ZD,NT,IWK(JWIPT),WK,WK(JWWPD))
      END IF
C
C Interpolate to get ZI values.
C
      ITPV=0
      JIG0MX=0
      JIG1MN=NXI*NYI+1
      NNGP=NT+2*NL
C
      DO 105 JNGP=1,NNGP
C
        ITI=JNGP
        IF (JNGP.LE.NT) GO TO 101
        IL1=(JNGP-NT+1)/2
        IL2=(JNGP-NT+2)/2
        IF (IL2.GT.NL) IL2=1
        ITI=IL1*(NT+NL)+IL2
C
  101   JWNGP=JWNGP0+JNGP
        NGP0=IWK(JWNGP)
        IF (NGP0.EQ.0) GO TO 103
        JIG0MN=JIG0MX+1
        JIG0MX=JIG0MX+NGP0
C
        DO 102 JIGP=JIG0MN,JIG0MX
          JWIGP=JWIGP0+JIGP
          IZI=IWK(JWIGP)
          IYI=(IZI-1)/NXI+1
          IXI=IZI-NXI*(IYI-1)
          CALL IDPTIP(XD,YD,ZD,NT,IWK(JWIPT),NL,IWK(JWIPL),WK,
     +                        ITI,XI(IXI),YI(IYI),ZI(IXI,IYI))
  102   CONTINUE
C
  103   JWNGP=JWNGP0+2*NNGP+1-JNGP
        NGP1=IWK(JWNGP)
        IF (NGP1.EQ.0) GO TO 105
        JIG1MX=JIG1MN-1
        JIG1MN=JIG1MN-NGP1
C
        DO 104 JIGP=JIG1MN,JIG1MX
          JWIGP=JWIGP0+JIGP
          IZI=IWK(JWIGP)
          IYI=(IZI-1)/NXI+1
          IXI=IZI-NXI*(IYI-1)
          CALL IDPTIP(XD,YD,ZD,NT,IWK(JWIPT),NL,IWK(JWIPL),WK,
     +                        ITI,XI(IXI),YI(IYI),ZI(IXI,IYI))
  104   CONTINUE
C
  105 CONTINUE
C
C Done.
C
        RETURN
C
      END

File: ncarg2d/man/bivar/idsfft.m
.TH IDSFFT 3NCARG "November 1995" UNIX "NCAR GRAPHICS"
.na
.nh
.SH NAME
IDSFFT - Performs smooth surface fitting when the projections
of the data points in the X-Y plane are irregularly distributed
in the plane.
.SH SYNOPSIS
 CALL IDSFFT (MD, NDP, XD, YD, ZD, MREG, NREG, KREG,
.br
+ XREG, YREG, ZREG, IWK, WK)
.SH C-BINDING SYNOPSIS
#include <ncarg/ncargC.h>
.sp
void c_idsfft (int md, int ndp, float *xd, float *yd,
.br
float *zd, int mreg, int nreg, int kreg, float *xreg,
.br
float *yreg, float *zreg, int *iwk, float *wk)
.SH DESCRIPTION
.IP "MD" 12
(Integer, Input/output) - 
Mode of computation (must be 1, 
2, or 3).
.RS
.IP "1"
If this is the first call to this subroutine, or if the value 
of NDP has been changed from the previous call, or if
the contents of the XD or YD arrays have been
changed from the previous call.
.IP "2"
If the values of NDP and the XD, YD arrays are
unchanged from the previous call, but new values for 
XREG, YREG are being used. If MD=2 and NDP has been
changed since the previous call to IDSFFT, an error return 
occurs.
.IP "3"
If the values of NDP, MREG, NREG, XD, YD, XREG,
YREG are unchanged from the previous call, that is, if the 
only change on input to IDSFFT is in the ZD array. If
MD=3 and NDP, MREG or NREG has been changed since the
previous call to IDSFFT, an error return occurs.
.sp
Between the call with MD=2 or MD=3 and the preceding 
call, the IWK and WK work arrays should not be disturbed.
.RE
.IP "NDP" 12
(Integer, Input) - 
Number of random data points (must be 4 
or greater).
.IP "XD(NDP)" 12
(Real array, Input) - 
Array of dimension NDP containing
the X coordinates of the data points.
.IP "YD(NDP)" 12
(Real array, Input) - 
Array of dimension NDP containing
the Y coordinates of the data points.
.IP "ZD(NDP)" 12
(Real array, Input) - 
Array of dimension NDP containing
the Z coordinates of the data points.
.IP "MREG" 12
(Integer, Input) - 
Number of output grid points in the  
X-direction (must be 1 or greater).
.IP "NREG" 12
(Integer, Input) - 
Number of output grid points in the  
Y-direction (must be 1 or greater).
.IP "KREG" 12
(Integer, Input) - 
First dimension of ZREG 
as declared in the calling program. KREG
must be greater than or equal to MREG,
else an error return occurs.
.IP "XREG(MREG)" 12
(Real array, Input) - 
Array of dimension MREG containing 
the X coordinates of the output grid points.
.IP "YREG(NREG)" 12
(Real array, Input) - 
Array of dimension NREG containing 
the Y coordinates of the output grid points.
.IP "ZREG(KREG,NREG)" 12
(Real array, Output) - 
Real, two-dimensional array of 
dimension (KREG,NREG), storing the interpolated Z values 
at the output grid points.
.IP "IWK(*)" 12 
(Integer array, Workspace) - Integer work array 
of dimension at least 31 * NDP + MREG * NREG.
.IP "WK(*)" 12 
(Real array, Workspace) - 
Real work array of 
dimension at least 6 * NDP.
.SH ""
Inadequate work space IWK and WK may may cause incorrect results.
.sp
The data points must be distinct and their projections in the
X-Y plane must not be collinear; otherwise, an error return
occurs.
.SH C-BINDING DESCRIPTION
The C-binding argument descriptions are the same as the FORTRAN 
.SH EXAMPLES
To use IDSFFT routines, load the NCAR Graphics libraries ncarg,
ncarg_gks, and ncarg_c, preferably in that order.
.sp
Use the ncargex command to see the following relevant
examples: 
ccpcldm,
ccpfil,
ccplbam,
ccpllb,
ccpllc,
ccplll,
ccpllo,
ccpllp,
ccpllt,
ccpllw,
ccpnet,
ccppc,
ccppc1,
ccppc2,
ccppc3,
ccppc4,
ccprc,
ccpscam,
cidsfft,
fsfsgfa,
cbex01.
.SH ACCESS 
To use IDSFFT or c_idsfft, load the NCAR Graphics libraries ncarg, ncarg_gks, 
and ncarg_c, preferably in that order.
.SH MESSAGES
See the bivar man page for a description of all Bivar error
messages and/or informational messages.
.SH SEE ALSO
Online:
bivar, bivar_params, idbvip, idpltr, idgeti, idgetr, idseti, idsetr,
ncarg_cbind
.sp
Hardcopy:
NCAR Graphics Contouring and Mapping Tutorial
.SH ACKNOWLEDGMENTS
Bivar was written by Hiroshi Akima in August 1975 and rewritten
by him in late 1976.  In 1989, a new version of Bivar,
incorporating changes described in a Rocky Mountain Journal of
Mathematics was obtained from Dr. Akima, and included in NCAR
Graphics with his permission.  In 1995, Dave Kennison incorporated
the capability of doing linear interpolation and a different kind
of triangulation, put in a parameter access interface, and wrote a
routine to allow the triangulation to be plotted.
.SH COPYRIGHT
Copyright (C) 1987-2009
.br
University Corporation for Atmospheric Research
.br
The use of this Software is governed by a License Agreement.
